Add ozw support for single setpoint thermostat devices

marcelveldt opened the pull request:
## Proposed change
This PR fixes support for (basic/simple) Z-wave thermostat devices that have no support for setting a mode with just a single setpoint like radiator valves. A good example is the Danfoss Living Connect Z thermostat. This device was not detected in the discovery.


## Type of change

- [ ] Dependency upgrade
- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests


## Additional information

- This PR fixes or closes issue: fixes #36541 

## Checklist

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

marcelveldt committed:
add support for thermostat without mode support

marcelveldt committed:
fix climate mode off

MartinHjelmare commented on homeassistant/components/ozw/climate.py:
```diff
@@ -274,8 +277,14 @@ async def async_set_fan_mode(self, fan_mode):
+            _LOGGER.warning(
+                "Thermostat %s does not support setting a mode", self.entity_id
+            )
+            return HVAC_MODE_HEAT
```
This method should not return anything. Just `return`.

marcelveldt replied:
Oops, how sloppy of me. Will fix

MartinHjelmare commented on homeassistant/components/ozw/discovery.py:
```diff
@@ -131,6 +131,37 @@
+            },
+            "temperature": {
+                const.DISC_COMMAND_CLASS: (CommandClass.SENSOR_MULTILEVEL,),
+                const.DISC_INDEX: (1,),
```
The magic numbers should probably be enum constants in the library.

marcelveldt replied:
Yes, we need to address constants for the ValueIndexes but that needs some discussion as ValueIndex can be the same between CC's. So if we want to do it 100% right, we should create the ValueIndex Enum/Constants for each CommandClass.

There's already an issue open for that in the lib as a reminder ;-)

marcelveldt committed:
very small is None fix

too small to get it's own PR. Value chould be checked against None instead in the rare case a temperature is below 0 degrees

marcelveldt committed:
fix return value

MartinHjelmare reviewed (approved):
Great!

MartinHjelmare commented:
Can we add a test for this type of node?

alex3305 commented on homeassistant/components/ozw/climate.py:
```diff
@@ -174,7 +174,8 @@ def on_value_update(self):
     def hvac_mode(self):
         """Return hvac operation ie. heat, cool mode."""
         if not self.values.mode:
-            return None
```
Small typo `condidered` should be `considered`.

alex3305 commented on homeassistant/components/ozw/climate.py:
```diff
@@ -174,7 +174,8 @@ def on_value_update(self):
         if not self.values.mode:
-            return None
+            # Thermostat(valve) with no support for setting a mode is condidered heating-only
+            return HVAC_MODE_HEAT
```
Would this also fix #28834 and perhaps also an issue related to #34788 but for Z-Wave devices? As I still cannot modify the temperature of my climate devices with the old Z-Wave component.

marcelveldt replied:
correct

balloob commented on homeassistant/components/ozw/climate.py:
```diff
@@ -174,7 +174,8 @@ def on_value_update(self):
         """Return hvac operation ie. heat, cool mode."""
         if not self.values.mode:
-            return None
+            # Thermostat(valve) with no support for setting a mode is condidered heating-only
```
```suggestion
            # Thermostat(valve) with no support for setting a mode is considered heating-only
```

balloob commented on homeassistant/components/ozw/climate.py:
```diff
@@ -331,20 +343,21 @@ def _get_current_mode_setpoint_values(self) -> Tuple:
-            else:
-                # treat value as hvac preset
-                all_presets[val[VALUE_LABEL]] = val[VALUE_ID]
+        if not self.values.mode:
+            all_modes[HVAC_MODE_HEAT] = None
+        else:
+            # Z-Wave uses one list for both modes and presets.
+            # Iterate over all Z-Wave ThermostatModes and extract the hvac modes and presets.
+            for val in self.values.mode.value[VALUE_LIST]:
+                if val[VALUE_ID] in MODES_LIST:
+                    # treat value as hvac mode
+                    hass_mode = ZW_HVAC_MODE_MAPPINGS.get(val[VALUE_ID])
+                    all_modes[hass_mode] = val[VALUE_ID]
+                else:
+                    # treat value as hvac preset
+                    all_presets[val[VALUE_LABEL]] = val[VALUE_ID]
```
Can we turn this around from `if not … else` to `if … else` ? 
```suggestion
        if self.values.mode:
            # Z-Wave uses one list for both modes and presets.
            # Iterate over all Z-Wave ThermostatModes and extract the hvac modes and presets.
            for val in self.values.mode.value[VALUE_LIST]:
                if val[VALUE_ID] in MODES_LIST:
                    # treat value as hvac mode
                    hass_mode = ZW_HVAC_MODE_MAPPINGS.get(val[VALUE_ID])
                    all_modes[hass_mode] = val[VALUE_ID]
                else:
                    # treat value as hvac preset
                    all_presets[val[VALUE_LABEL]] = val[VALUE_ID]
        else:
            all_modes[HVAC_MODE_HEAT] = None
```

marcelveldt committed:
Update homeassistant/components/ozw/climate.py

Co-authored-by: Paulus Schoutsen <paulus@home-assistant.io>

marcelveldt committed:
Update homeassistant/components/ozw/climate.py

Co-authored-by: Paulus Schoutsen <paulus@home-assistant.io>

marcelveldt committed:
add test

marcelveldt commented:
> Can we add a test for this type of node?

done!

marcelveldt committed:
update tests

code coverage

MartinHjelmare commented:
Maybe try rebasing to have the github ci run again and hopefully solve the problem.

marcelveldt commented:
great, rebasing did not work out great. will have to fix that

marcelveldt commented:
more ideas for the code coverage ?

MartinHjelmare commented on homeassistant/components/ozw/climate.py:
```diff
@@ -220,6 +221,8 @@ def hvac_action(self):
         """Return preset operation ie. eco, away."""
         # A Zwave mode that can not be translated to a hass mode is considered a preset
+        if not self.values.mode:
+            return None
```
This is the only non covered line for the diff.

MartinHjelmare reviewed (approved):
I think coverage is ok. Codecov doesn't report it correctly for this branch HEAD.
